### ISSUE TITLE
The special case for Filter assumes that DecodeParms are always present

### ISSUE THREAD

**Issue by bdoubrov** [pdf-association/arlington-pdf-model#62]
Currently **Stream** and other stream-based objects define the special case condition 
```
fn:Eval(fn:ArrayLength(DecodeParms) == fn:ArrayLength(Filter))
```
for both **Filter** and **DecodeParms** keys. As a result, the following perfectly valid use case:
```
<<
/Filter [/FlateDecode]
% no DecodeParms
...
>>
stream
...
endstream
```
is reporting an error because of the special **Filter** condition.

There are two ways to fix this:
1) either remove completely the special case for **Filter**
2) or replace **Filter** special case condition by 
```
fn:Eval(fn:IsPresent(DecodeParms,fn:ArrayLength(DecodeParms) == fn:ArrayLength(Filter)))
```

The same applies to **FFilter** and **FDecodeParms** keys.

**Comment by faceless2** [comment]
This caught me too the first time around. If `DecodeParms` is not present, then `fn:ArrayLength(DecodeParms)` doesn't evaluate to zero, it evaluates to _undefined_, which makes the whole expression undefined. See https://github.com/pdf-association/arlington-pdf-model/issues/30

**Comment by petervwyatt** [comment]
As @faceless2 says...

**Comment by petervwyatt** [comment]
See also #47

**Comment by bdoubrov** [comment]
Not sure if this is the best place to continue the discussion on the formal logic, or reopen #47. But as this issue is still open, I leave the comment here.

While the general use of _undefined_ for missing properties is fine, I see some issues that the current Arlington boolean logic is not aligned with ECMAScript that also supports _undefined_. See https://tc39.es/ecma262/multipage/global-object.html#sec-undefined 

The key difference is that the expression `(a == undefined)` is always _false_ in JavaScript, while is undefined in Arlington. But there are other strange things like `(false && undefined)` is also _undefined_ in Arlington, which prevents the use of [short circuit evaluation](https://en.wikipedia.org/wiki/Short-circuit_evaluation). 

Is there a fundamental problem in aligning Arlington boolean logic with ECMAScript one?

**Comment by petervwyatt** [comment]
Yes - I'm trying to use predicate logic, NOT functional programming logic.
